File: infra/src/main/java/mixter/infra/repositories/InMemoryFollowerRepository.java
package mixter.infra.repositories;

import mixter.domain.core.subscription.FollowerRepository;
import mixter.domain.identity.UserId;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class InMemoryFollowerRepository implements FollowerRepository {
    Map<UserId, Set<UserId>> followers = new HashMap<>();

    @Override
    public Set<UserId> getFollowers(UserId followee) {
        return followers.getOrDefault(followee, emptySet());
    }

    private Set<UserId> emptySet() {
        return new HashSet<>();
    }

    @Override
    public void saveFollower(UserId followee, UserId follower) {
        Set<UserId> userFollowers = this.followers.getOrDefault(followee, emptySet());
        userFollowers.add(follower);
        this.followers.put(followee, userFollowers);
    }

    @Override
    public void removeFollower(UserId followee, UserId follower) {
        Set<UserId> userFollowers = this.followers.getOrDefault(followee, emptySet());
        userFollowers.remove(follower);
        this.followers.put(followee, userFollowers);
    }
}

File: infra/src/main/scala/mixter/infra/repositories/InMemoryFollowerRepository.scala
package mixter.infra.repositories

import mixter.domain.identity.UserId
import mixter.domain.subscription.FollowerRepository

class InMemoryFollowerRepository extends FollowerRepository {
  private var followers = Map.empty[UserId, Set[UserId]].withDefaultValue(Set.empty)

  override def getFollowers(authorId: UserId): Set[UserId] = followers(authorId)

  override def saveFollower(followee: UserId, follower: UserId): Unit =
    followers+= followee -> (followers(followee)+follower)

  override def removeFollower(followee: UserId, follower: UserId): Unit =
    followers+= followee -> (followers(followee)-follower)
}
